tensorflow/compiler/jit/cluster_scoping_pass.cc

```diff
@@ -0,0 +1,148 @@
+  ClusterScopingPassImpl(Graph* graph,
+                         OptimizerOptions::GlobalJitLevel global_jit_level)
+      : graph_(graph),
+        global_jit_level_(global_jit_level),
+        unique_scope_id_(0) {}
+
+  Status Run();
+
+ private:
+  Status ScopingForPipelineStages();
+
+  size_t GetUniqueScopeId() { return unique_scope_id_++; }
+
+  void AddScopeToAllPredecessors(Node* start);
+
+  void AddScopeToAllSuccessors(Node* start);
+
+ private:
+  Graph* graph_;
+  OptimizerOptions::GlobalJitLevel global_jit_level_;
+  size_t unique_scope_id_;
+};
+
+absl::optional<string> GetXlaAutoJitScope(Node* node) {
+  string scope;
+  if (GetNodeAttr(node->attrs(), kXlaAutoJitScopeAttr, &scope).ok()) {
+    return scope;
+  }
+
+  return absl::nullopt;
+}
+
+void SetXlaAutoJitScope(Node* node, StringPiece scope) {
+  node->AddAttr(kXlaAutoJitScopeAttr, scope);
+}
+
+// NB! We append a new scope as suffix to the XlaAutoJitScope attribute instead
+// of overriding the old value.  In this way, we respect the original scopes.
+// In other words, appending X to Y creates the conjunction of the scopes X
+// and Y (i.e, X & Y in effect).
+void AddOrAppendXlaAutoJitScope(Node* node, absl::string_view suffix) {
+  string updated_scope;
+  absl::optional<string> cur_scope = GetXlaAutoJitScope(node);
+  if (cur_scope == absl::nullopt) {
+    updated_scope = std::string(suffix);
+  } else {
+    updated_scope = absl::StrCat(cur_scope.value(), "&", suffix);
+  }
+  SetXlaAutoJitScope(node, updated_scope);
+}
+
+void ClusterScopingPassImpl::AddScopeToAllPredecessors(Node* start) {
+  const string unique_suffix = absl::StrCat("_", GetUniqueScopeId());
+
+  std::vector<Node*> starts;
+  starts.push_back(start);
+  auto enter = [&](Node* n) { AddOrAppendXlaAutoJitScope(n, unique_suffix); };
+  ReverseDFSFrom(*graph_, starts, enter, /*leave=*/nullptr,
+                 /*stable_comparator=*/NodeComparatorName());
+}
+
+void ClusterScopingPassImpl::AddScopeToAllSuccessors(Node* start) {
+  const string unique_suffix = absl::StrCat("_", GetUniqueScopeId());
+
+  std::vector<Node*> starts;
+  starts.push_back(start);
+  auto enter = [&](Node* n) { AddOrAppendXlaAutoJitScope(n, unique_suffix); };
+  auto not_back_edge = [](const Edge& edge) -> bool {
+    return !edge.src()->IsNextIteration();
+  };
+  DFSFrom(*graph_, starts, enter, /*leave=*/nullptr,
+          /*stable_comparator=*/NodeComparatorName(),
+          /*edge_filter=*/not_back_edge);
+}
+
+Status ClusterScopingPassImpl::ScopingForPipelineStages() {
+  for (Node* n : graph_->nodes()) {
+    DCHECK(n);
+    if (n->type_string() == "Unstage") {
+      AddScopeToAllSuccessors(n);
+    }
+    if (n->type_string() == "Stage") {
+      AddScopeToAllPredecessors(n);
+    }
+  }
+
+  return Status::OK();
+}
+
+Status ClusterScopingPassImpl::Run() {
+  if (global_jit_level_ == OptimizerOptions::OFF) {
+    return Status::OK();
+  }
+
+  // This preserves the parallelism between pipeline stages.  For example,
```

Comment:
This comment should live in `ScopingForPipelineStages` I think.

Reply:
Agreed.